vortex-ipc/src/messages/decoder.rs

```diff
@@ -0,0 +1,289 @@
+                    bytes.reserve_aligned(*buffers_length, self.alignment);
+                    return Ok(PollRead::NeedMore(*buffers_length));
+                }
+
+                // SAFETY: we've already validated the header
+                let msg = unsafe { root_unchecked::<fb::Message>(header.as_ref()) };
+                let array_data_msg = msg
+                    .header_as_array_data()
+                    .vortex_expect("array data header");
+                let array_msg = array_data_msg
+                    .array()
+                    .ok_or_else(|| vortex_err!("array data message missing array"))?;
+
+                let buffers = array_data_msg
+                    .buffers()
+                    .unwrap_or_default()
+                    .iter()
+                    .map(|buffer_msg| {
+                        let buffer_len = usize::try_from(buffer_msg.length())
+                            .vortex_expect("buffer length is too large for usize");
+                        let buffer = bytes.split_to_aligned(buffer_len, self.alignment);
+                        let _padding = bytes.split_to(buffer_msg.padding() as usize);
+                        Buffer::from(buffer.freeze())
+                    })
+                    .collect_vec();
+
+                let row_count = usize::try_from(array_data_msg.row_count())
+                    .map_err(|_| vortex_err!("row count is too large for usize"))?;
+
+                let msg = DecoderMessage::Array(ArrayParts {
+                    row_count,
+                    array_flatbuffer: header.clone(),
+                    array_flatbuffer_loc: array_msg._tab.loc(),
+                    buffers,
+                });
+                self.state = State::Length;
+                Ok(PollRead::Some(msg))
+            }
+        }
+    }
+}
+
+trait BytesMutAlignedSplit {
+    /// Advance the buffer until the start is aligned, and then reserve the requested capacity.
+    /// If the buffer is non-empty, this function does nothing.
+    fn reserve_aligned(&mut self, additional: usize, align: usize);
+
+    /// Splits the buffer at the given index, ensuring the returned BytesMut is aligned
+    /// as requested. This may involve a copy. Note that in practice, since we reserve_aligned
+    /// before calling this, the copy should be very are.
+    fn split_to_aligned(&mut self, at: usize, align: usize) -> BytesMut;
+}
+
+impl BytesMutAlignedSplit for BytesMut {
+    fn reserve_aligned(&mut self, additional: usize, align: usize) {
+        if !self.is_empty() {
+            // If the buffer isn't empty, we shouldn't try to copy it. It could be huge and this
+            // is just an optimization to avoid the copy in `split_to_aligned`.
+            return;
+        }
+
+        // Reserve up to the worst-cast alignment
+        self.reserve(additional + align);
+        let padding = self.as_ptr().align_offset(align);
+        unsafe { self.set_len(padding) };
```

Comment:
don't you want to account for the case where len != 0 before this method is called?